File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.diameter = 0

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.dfs(root, 0)
        return self.diameter

    def dfs(self, root, curr_depth):
        if root is None:
            return curr_depth

        left_d = self.dfs(root.left, curr_depth)
        right_d = self.dfs(root.right, curr_depth)

        self.diameter = max(self.diameter, left_d + right_d)
        print(self.diameter)

        return 1 + max(left_d, right_d)

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int res = -1;
        dfs(root, k, res);
        return res;
    }

private:
    int dfs(TreeNode *node, int k, int &res) {
        if (node == nullptr) return 0;

        int num_nodes_bigger_then = dfs(node->left, k, res);
        if (num_nodes_bigger_then + 1 == k) res = node->val;

        num_nodes_bigger_then += dfs(node->right, k - (num_nodes_bigger_then + 1), res);

        return num_nodes_bigger_then + 1;
    }
};
